File: 0050.Pow(x,n)/0050-Pow(x,n).py

from functools import lru_cache
class Solution:
    def myPow(self, x: float, n: int) -> float:
        def square(x):
            return x*x
        @lru_cache(maxsize=32)
        def h(x,n):
            if n == 0:
                return 1
            if n & 1 == 0:
                return square( h(x,n//2) )
            else:
                return x * h(x, n-1)
        return h(x,n) if n > 0 else 1/h(x,-n)
            
        

File: 0050.Pow(x,n)/0050-Pow(x,n).rs
impl Solution {
    pub fn my_pow(x: f64, n: i32) -> f64 {
        fn h(x: f64, n: i64) -> f64 {
            if n == 0{
                return 1.0;
            }
            if n & 1 == 0 {
                return h(x,n/2) * h(x, n/2);
            } else {
                return x * h(x,n-1);
            }
        }
        let n = n as i64;
        if n > 0 { h(x,n) } else { 1.0/h(x,-n) }
    }
}